feat: Log FTT2 loss + Fourier Correlation loss

OpheliaMiralles opened the pull request:
## Description

<!-- Provide a brief summary of the changes introduced in this pull request. -->

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

<!-- Link the Issue number this change addresses, ideally in one of the "magic format" such as Closes #XYZ -->

<!-- Alternatively, explain the motivation behind the changes and the context in which they are being made. -->
new loss function

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [x] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [x] I have tested the changes on a single GPU
-   [x] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code
-   [ ] If the new feature introduces modifications at the config level, I have made sure to update Pydantic Schemas and default configs accordingly

<!-- In case this affects the model sharding or other specific components please describe these here. -->

### Dependencies

-   [x] I have ensured that the code is still pip-installable after the changes and runs
-   [x] I have tested that new dependencies themselves are pip-installable.
-   [x] I have not introduced new dependencies in the inference portion of the pipeline

<!-- List any new dependencies that are required for this change and the justification to add them. -->

### Documentation

-   [x] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [x] I have added comments to my code, particularly in hard-to-understand areas

<!-- Describe any major updates to the documentation -->

## Additional Notes

<!-- Include any additional information, caveats, or considerations that the reviewer should be aware of. -->


<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--148.org.readthedocs.build/en/148/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--148.org.readthedocs.build/en/148/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--148.org.readthedocs.build/en/148/

<!-- readthedocs-preview anemoi-models end -->

mchantry commented:
Exciting contribution, thanks.

mchantry commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,64 @@
+    return (10 * log10(ratio)) ** 2
+
+
+class LogSpectralDistance(FunctionalWeightedLoss):
```
I wonder if a different name might be better, more precise that this uses 2DFFTs for the spectra.
It's possible we may want to add a Spherical harmonic spectral distance later, and both could be interpreted as a spectral distance.

OpheliaMiralles replied:
Which name would you suggest? I am also wondering whether this current implementation is correct (that is why I left it in draft mode) because anemoi-datasets uses flattened spatial coordinates. What I had in mind is computing the 2D fourier transform on time/batch sequences of 2D fields. I have another implementation, however, it takes as input of the loss a field shape and reshape the tensor before computing the loss, e.g. it would not work with unstructured grids. So I don't know whether to drop it or update the PR with the 2D version and if the x/y dim are not provided, it would raise an error anyway.

mchantry commented:
Sorry for the slow re-review Ophelia.
I think this will work for rectilinear grids. Do you have a thought on how we could ensure that it is only used in this context, and not misused?
Would you be able to provide an update to the docs please.

OpheliaMiralles commented:
> Sorry for the slow re-review Ophelia. I think this will work for rectilinear grids. Do you have a thought on how we could ensure that it is only used in this context, and not misused? Would you be able to provide an update to the docs please.

No worries, I was waiting for a signal to start working on this again. I might have some time next week or the following.

OpheliaMiralles commented:
> Sorry for the slow re-review Ophelia. I think this will work for rectilinear grids. Do you have a thought on how we could ensure that it is only used in this context, and not misused? Would you be able to provide an update to the docs please.

Updated the doc and added a warning

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+    real_output: torch.Tensor,
+    dims: tuple[int, int],
+) -> tuple[torch.Tensor, torch.Tensor]:
+    assert dims[0] * dims[1] == real_output.shape[2]
```
Can you add a message to this assert, so if it is triggered, a useful message is shown

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+class LogFFT2Distance(FunctionalLoss):
+    r"""The log spectral distance is used to compute the difference between spectra of two fields.
+
+    It is also called log spectral distorsion.
```
`Distortion`?

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+    return power_spectra_real, power_spectra_fake
+
+
+def log_rfft2_distance(fake_output: torch.Tensor, real_output: torch.Tensor, dims: tuple[int, int]) -> torch.Tensor:
```
Why is the prediction being called `fake_output`?

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+    power_spectra_fake = torch.fft.rfft2(fake_output.reshape(dims_total), dim=(-2, -3))
+    return power_spectra_real, power_spectra_fake
+
+
+def log_rfft2_distance(fake_output: torch.Tensor, real_output: torch.Tensor, dims: tuple[int, int]) -> torch.Tensor:
+    power_spectra_real, power_spectra_fake = get_spectra(fake_output, real_output, dims)
+    epsilon = torch.finfo(real_output.dtype).eps  # Small epsilon to avoid division by zero
```
Please add a docstring to this function

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+        )
+        self.x_dim = x_dim
+        self.y_dim = y_dim
+        LOGGER.warning("Please use x_dim and y_dim such that field_shape=(x_dim, y_dim).")
```
Rather then raise a warning here, can you add a check and exception in the `calculate` function?

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+        without_scalars: list[str] | list[int] | None = None,
+    ) -> torch.Tensor:
+        result = super().forward(pred, target, squash, scalar_indices=scalar_indices, without_scalars=without_scalars)
+        return torch.sqrt(torch.mean(result))
```
In the case of `squash==True`, I believe the dimensions will have already been reduced and averaged. Does the `sqrt(` belong in the `calculate_difference`?

HCookie commented on training/src/anemoi/training/losses/spatial.py:
```diff
@@ -0,0 +1,155 @@
+        )
+        self.x_dim = x_dim
+        self.y_dim = y_dim
+        LOGGER.warning("Please use x_dim and y_dim such that field_shape=(x_dim, y_dim).")
```
Same note as above

HCookie commented on training/src/anemoi/training/losses/loss.py:
```diff
@@ -69,7 +69,7 @@ def get_loss_function(
+    if loss_config.get("x_dim", None) is not None and loss_config.get("y_dim", None) is not None:
         kwargs["x_dim"] = config.x_dim
         kwargs["y_dim"] = config.y_dim
 
```
If they are in the loss config, why pull them out into the kwargs, they are used to instantiate?

OpheliaMiralles replied:
then I guess we don't need it at all indeed, could you please remove those lines? not working this afternoon but would love to see this merged
